File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.java
class Solution {

    public int[] searchRange(int[] nums, int target) {

        return new int[]{
            findFirst(nums, target),
            findLast(nums, target)
        };
    }

    private int findFirst(int[] nums, int target) {

        int l = 0;
        int r = nums.length - 1;
        int ans = -1;

        while (l <= r) {

            int m = (l + r) / 2;

            if (nums[m] == target) {
                ans = m;
                r = m - 1;       // search left
            }
            else if (nums[m] < target) {
                l = m + 1;
            }
            else {
                r = m - 1;
            }
        }

        return ans;
    }

    private int findLast(int[] nums, int target) {

        int l = 0;
        int r = nums.length - 1;
        int ans = -1;

        while (l <= r) {

            int m = (l + r) / 2;

            if (nums[m] == target) {
                ans = m;
                l = m + 1;       // search right
            }
            else if (nums[m] < target) {
                l = m + 1;
            }
            else {
                r = m - 1;
            }
        }

        return ans;
    }
}

File: 34-find-first-and-last-position-of-element-in-sorted-array/find-first-and-last-position-of-element-in-sorted-array.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        def findfirst():
            l=0
            r=len(nums)-1
            ans=-1
            while(l<=r):
                m=(l+r)//2

                if nums[m] == target:
                    ans=m
                    r=m-1 #search left
                elif nums[m]<target:
                    l = m+1
                else:
                    r = m-1
            return ans
            
        def findlast():
            l=0
            r=len(nums)-1
            ans=-1
            while(l<=r):
                m=(l+r)//2

                if nums[m] == target:
                    ans=m
                    l=m+1 #search right
                elif nums[m]<target:
                    l = m+1
                else:
                    r = m-1
            return ans
        return [findfirst(),findlast()]